internal/jujuapi/http_proxy_integration_test.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2024 Canonical.
+
+package jujuapi_test
+
+import (
+	"context"
+	"encoding/pem"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"strings"
+
+	"github.com/juju/charm/v12"
+	"github.com/juju/charm/v12/resource"
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/api/client/charms"
+	"github.com/juju/juju/api/client/resources"
+	"github.com/juju/juju/core/network"
+	jujuparams "github.com/juju/juju/rpc/params"
+	"github.com/juju/juju/testcharms"
+	"github.com/juju/juju/testing/factory"
+	"github.com/juju/names/v5"
+	"github.com/juju/version/v2"
+	gc "gopkg.in/check.v1"
+
+	"github.com/canonical/jimm/v3/internal/jujuapi"
+)
+
+type httpProxySuite struct {
+	websocketSuite
+}
+
+var _ = gc.Suite(&httpProxySuite{})
+
+func (s *httpProxySuite) TestHTTPAuthenticate(c *gc.C) {
+	ctx := context.Background()
+	httpProxier := jujuapi.HTTPProxier(s.JIMM)
+
+	// good
+	req, err := http.NewRequest("POST", fmt.Sprintf("/%s/charms", s.Model.UUID.String), nil)
+	c.Assert(err, gc.IsNil)
+	token, err := s.JIMM.OAuthAuthenticator.MintSessionToken(s.AdminUser.Name)
+	c.Assert(err, gc.IsNil)
+	req.SetBasicAuth("", token)
+	c.Assert(err, gc.IsNil)
+	err = httpProxier.Authenticate(ctx, nil, req)
+	c.Assert(err, gc.IsNil)
+
+	// missing auth
+	req, err = http.NewRequest("POST", fmt.Sprintf("/%s/charms", s.Model.UUID.String), nil)
+	c.Assert(err, gc.IsNil)
+	err = httpProxier.Authenticate(ctx, nil, req)
+	c.Assert(err, gc.ErrorMatches, "authentication missing")
+
+	// wrong user
+	req, err = http.NewRequest("POST", fmt.Sprintf("/%s/charms", s.Model.UUID.String), nil)
+	c.Assert(err, gc.IsNil)
+	token, err = s.JIMM.OAuthAuthenticator.MintSessionToken("test-user")
+	c.Assert(err, gc.IsNil)
+	req.SetBasicAuth("", token)
+	c.Assert(err, gc.IsNil)
+	err = httpProxier.Authenticate(ctx, nil, req)
+	c.Assert(err, gc.ErrorMatches, "unauthorized")
+}
+
+func (s *httpProxySuite) TestCharmHTTPServe(c *gc.C) {
+	ctx := context.Background()
+	httpProxier := jujuapi.HTTPProxier(s.JIMM)
+	expectU, expectP, err := s.JIMM.GetCredentialStore().GetControllerCredentials(ctx, s.Model.Controller.Name)
+	c.Assert(err, gc.IsNil)
+	// we expect the controller to respond with TLS
+	fakeController := httptest.NewTLSServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
If we're using a fake controller then we shouldn't be using the websocketSuite (that spins up a real Juju controller), we can use one of the simpler suites that just spins up JIMM.